### DISCUSSION TITLE
Introduce Support of Latest Java Versions

### DISCUSSION THREAD

**Discussion by amitjoy** [shannah/jdeploy#78] [General]
There are new Java versions available - 18 and 19 that introduce lots of important features which the users could also leverage. I believe, it would be better to introduce the support of these versions too.

**Comment by shannah** [comment]
I think this may just require a change in the GUI to give you the option for newer versions.
If you modify your package.json file manually, and set the javaVersion property to '18', or '19', it should work already.  Can you let me know if that doesn't work for you?

(Whether it works or not, I'll keep this issue opened until I add this in the GUI)

**Comment by Oliver-Loeffler** [comment]
I've tried to use Java 20 but it stuck with Java 17 (and JavaFX 19).
The GUI approach is nice.

**Comment by shannah** [comment]
You should be able to specify the Java version manually in the package.json file.  Just set the `javaVersion` property of the "jdeploy" object.  E.g.

```
{
   ...
   "jdeploy": {
        ...
        "javaVersion": "20",
        ...
  }
  ...
}
```

**Comment by shannah** [comment]
Actually, just tried it, and the GUI does have some problems handling this.  While it currently works to just set this manually, the GUI will give you problems when you try to edit it - it will overwrite it back to one of the supported values on save.

I have just made a change to fix this, and it will be part of 4.0.22 which will be available within the next hour or so.

**Comment by Oliver-Loeffler** [comment]
Thanks for testing, I also analysed it and found that the Zulu JDK is used with JavaFX 20. All perfect, I was wrong with that.
Well I also learned that the info page of Scene Builder requires more care, it is just misleading. 

![grafik](https://user-images.githubusercontent.com/22102800/227791737-67eef534-d4d0-4162-9376-f4fe4e6ea9ed.png)

**Comment by shannah** [comment]
If you really want to work with the bleeding edge of javaFX, you can add a javafxVersion property to specify a specific version. 
E.g.

```
{
   ...
   "jdeploy": {
        ...
        "javafxVersion": "21-ea+5",
        ...
  }
  ...
}
```

When you specify this option it will cause the installer to not use the javafx that is bundled with zulu (it will use the regular zulu release without javafx), and will download the javafx dependencies directly from maven.

This is a little bit experimental at this point, however, so I would only use this for tinkering.

**Reply by Oliver-Loeffler** [reply]
Thanks for pointing this out, yes, I will definitely make use of this as soon as Scene Builder 20 is released. We can than integrate all the new stuff with latest JavaFX. For that I will thankfully look into the JavaFX starte project, think this will be a great help.

**Comment by shannah** [comment]
I think it is cool that you published scenebuilder leading edge this way.  I just tried installing it on my system and it seems to be working ok.  I hadn't yet tried out the "leading edge" version.

**Comment by Oliver-Loeffler** [comment]
The actual "release" version will follow. I'm just working on 17/18/19, get them all ready. The "LE" will not interfere, it uses its own repository, its own settings etc.
